### PR TITLE
[WIP] PPMS token proof of concept.

### PR DESCRIPTION
**Pull request by SimonSchama**
## Description of change
This is a proof of concept demonstrating an approach to acquiring the bearer tokens that will be necessary at an undetermined point in the future to authenticate to ppms. Currently this authentication is only enabled for the /CareSites endpoint in the dev and nonprod ppms. For this code to work properly, the environment specific authority, tenant, resource, client id, and client secret need to be added into settings.local.yml. I have this information/a set of credentials for dev (also works for nonprod), someone on the DS team should probably request this information for production when they decide to roll these changes out to production.

Due to incompatibilities with the gems we're using, I opted to do this without a library. Due to the simplicity of the use case, I did not bother creating a client class to wrap the connection to the azure active directory authority. I also didn't put the key for the cached redis token into any namespace. That can be adjusted if desired. Other potential improvements would be to refactor the method into a Faraday middleware to make it easier to hook into all the existing ppms calls or to improve error handling to potentially allow this code to be deployed prior to authentication becoming possible in a production environment. 

Overall I'm looking for input on the approach taken and when and how it should be integrated with the rest of the website given that it's currently unknown when the token will be available/required for production (not for a few months at least).

## Testing done
Tested using rails console in a review instance environment against dev ppms/azure active directory.

#### Unique to this PR

- [x] Successfully obtains token and connects to authenticated ppms endpoint.
- [ ] Properly hooked into all ppms calls.

#### Applies to all PRs

- [ ] Appropriate logging
- [ ] Swagger docs have been updated, if applicable
- [ ] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [ ] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

**Comment by rtluu** [comment]
@department-of-veterans-affairs/backend-review-group could one of you take a look at this PR?

File: lib/facilities/ppms_client.rb [hunk 1]
```diff
@@ -44,6 +44,26 @@ def provider_services(identifier)
       response.body
     end
 
+    def fetch_token
+      token = Redis.current.get('ppms-token')
+      return token if token
+      azure_client = Faraday.new("https://#{Settings.ppms.authority}")
```
Comment (kreek) [inline lib/facilities/ppms_client.rb]: Realize it's a simple call to fetch a token for use in the PPMS client but we end up with custom Faraday code that uses a different pattern than all our other http calls so I think it implement a client of its own. A custom Faraday call also means it doesn't automatically get wired up to our monitoring and default error handling. What would happen if this call failed? A spec using a VCR cassette that returns a 500 would be one way to find out 😄

File: lib/facilities/ppms_client.rb [hunk 2]
```diff
@@ -44,6 +44,26 @@ def provider_services(identifier)
...
+      auth_body = { 'encoding': 'utf8', 'api-version': '1.0',
+                    'client_id': Settings.ppms.client_id, 'client_secret': Settings.ppms.client_secret,
+                    'grant_type': 'client_credentials', 'resource': Settings.ppms.resource }
+      token_response = azure_client.post "/#{Settings.ppms.tenant}/oauth2/token", auth_body
+      token_hash = JSON.parse(token_response.body)
```
Comment (kreek) [inline lib/facilities/ppms_client.rb]: ... and if this had its own client implementation you can have Faraday middleware [parse the body](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/salesforce/configuration.rb#L13)

**Comment by johnpaulashenfelter** [comment]
@SimonSchama is this still in process?

**Comment by kfrz** [comment]
Closing as this is not linked to any ticket and has staled.

**Pull request closed without merging**